core/base/src/TSystem.cxx

```diff
@@ -2842,7 +2845,50 @@ int TSystem::CompileMacro(const char *filename, Option_t *opt,
       withInfo = strchr(opt, 's') == nullptr;
       verbose = strchr(opt, 'v') != nullptr;
       internalDebug = strchr(opt, 'd') != nullptr;
+      compileIfDifferent = strchr(opt,'h') != nullptr;
    }
+
+   // ======= Get the right file names for the dictionary and the shared library
+   TString expFileName(filename);
+   ExpandPathName( expFileName );
+   expFileName = gSystem->UnixPathName(expFileName);
+
+   std::string hash_symbol_name = "__ROOT_Internal_CompileMacro_";
+   if (compileIfDifferent) {
+      std::ifstream inputFile(expFileName);
+      if (!inputFile) {
+         ::Error("ACLiC", "The input macro file '%s' could not be opened.", expFileName.Data());
+         return 0;
+      }
+      std::string fileContent((std::istreambuf_iterator<char>(inputFile)), (std::istreambuf_iterator<char>()));
+      inputFile.close();
+      fileContent += expFileName;
+      auto hash = std::hash<std::string>{}(fileContent);
+      hash_symbol_name += std::to_string(hash);
+
+      // We need to check for the existence of the symbol from within libCling. This is because that library
+      // is loaded specifying RTLD_LOCAL and therefore hiding all symbols from the process as well as all the
+      // symbols of the libraries it loads. Therefore, we jit a helper for that.
+      static void* (*dlsym_jit)(const char *) = nullptr;
+      if (!dlsym_jit) {
+         const std::string dlsym_jit_name("__ROOT_Internal_CompileMacro_lsym_jit");
+         std::string dlsym_jit_code = "void *";
+         dlsym_jit_code +=
+            dlsym_jit_name + "(const char* hash_symbol_name) {return dlsym(RTLD_DEFAULT, hash_symbol_name);} ";
```

Comment:
`dlsym` is (was?) not portable to Windows.  See `TROOT.cxx` `TCling.cxx` `TProtoClass.cxx` and `TClass.cxx` (where `TROOT.cxx` seems to be the more recent one.  (and yes we could use a factorization of those 4 instances).